### DISCUSSION TITLE
Agreement ID on Provider side

### DISCUSSION THREAD

**Discussion by johnnychoque** [eclipse-edc/Connector#4277] [Q&A]
Hello all,

Following the steps in the "Transfer Samples - consumer pull" repository, I have managed to get the Contract Agreement ID from the Consumer side using the API `GET /management/v2/contractnegotiations/<<NegotiationID>>`. To integrate the connector with another system, I need to get the Contract Agreement ID from the Provider side. As each connector manages its state machine, I understand that the Negotiation ID of the Consumer is different from the Provider. How can I get the Negotiation ID from the Provider side? 

I am using EDC 0.7.0 and in EDC Management API I have seen that there is the API `POST /v2/contractnegotiations/request`. According to the documentation, this API returns all contract negotiations according to a query. 
[https://app.swaggerhub.com/apis/eclipse-edc-bot/management-api/0.7.0#/Contract%20Negotiation/queryNegotiations](https://app.swaggerhub.com/apis/eclipse-edc-bot/management-api/0.7.0#/Contract%20Negotiation/queryNegotiations)

```bash
curl 'http://localhost:19193/v2/contractnegotiations/request' \
--header 'Content-Type: application/json' \
--data-raw '{
  "@context": {
    "@vocab": "https://w3id.org/edc/v0.0.1/ns/"
  },
  "@type": "QuerySpec",
  "offset": 5,
  "limit": 10,
  "sortOrder": "DESC",
  "sortField": "fieldName",
  "filterExpression": []
}'
```
I have tried to use it in the Consumer and Provider but in both it returns "**Error 405 HTTP method POST is not supported by this URL**". Is this API not implemented yet?

I would appreciate if you could give me some idea on how to get the Contract Agreement ID on the Provider side.

Regards,

Johnny

**Comment by ndr-brt** [comment]
The contract agreement id is unique, this means that has the same value both on provider and consumer side.

Given the agreement id, you can get the negotiation using this API:
https://eclipse-edc.github.io/Connector/openapi/management-api/0.7.0/#/Contract%20Agreement/getNegotiationByAgreementId

please note that the swaggerhub publication has been deprecated, now we're publishing the openapi ui on github pages:
https://eclipse-edc.github.io/Connector/openapi/management-api

**Reply by johnnychoque** [reply]
Hi @ndr-brt,

Thanks for clearing up the contract agreement id and the management-api repository.

In the consumer pull sample, the consumer app gets the contract negotiation id inside the response to the negotiation initiation request (`POST /management/v2/contractnegotiations`). However, the provider app does not receive any message so it does not know the negotiation id. Both values, the negotiation id and the agreement id, are only known in the respective connectors (see figure below).

![Negotiation](https://github.com/eclipse-edc/Connector/assets/17832942/f22e6423-9fa9-4d24-843a-82ba2f78da29)

Although on the Consumer side, the contract agreement id is known and this value is unique, considering that the Consumer and Provider are isolated environments, what I would need is an API that allows the provider app to get the contract agreement id on the Provider side. 

I have managed to get what I need using this API:
[https://eclipse-edc.github.io/Connector/openapi/management-api/0.7.0/#/Contract%20Negotiation/queryNegotiations](https://eclipse-edc.github.io/Connector/openapi/management-api/0.7.0/#/Contract%20Negotiation/queryNegotiations)

But it returns all the contract negotiations that exist so far and I can't filter the current negotiation based on any data that the provider app knows (e.g. asset, policy, or offer).

Maybe an alternative is to create some component that interacts with the state machine of the negotiation process so that the Provider connector sends to the Provider app the contract negotiation id when it receives the dataspace protocol api `POST /protocol/negotiations/request`. what do you think?

Regards,

Johnny

**Reply by ndr-brt** [reply]
I'm not sure to understand why, from the sample point of view, the provider would need to know the contract agreement id.

Anyway, you could take a look at the eventing system, it would be just enough to register an `EventSubscriber` on the `ContractNegotiationFinalized` type of event (that contains the agreement) so the agreement can be stored/sent/anything.

https://eclipse-edc.github.io/docs/#/submodule/Connector/docs/developer/events
